### ISSUE TITLE
RNN FLIP

### ISSUE THREAD

**Issue by cgarciae** [google/flax#2396]
FLIP doc migrated to #2585

This flip proposes adding the `RNN` and `Bidirectional` classes that would help simplify code and handle some of the edge cases regarding padded bidirectional RNNs.

### Proposal Summary

**RNN**
```python
cell = nn.LSTMCell()
# Encodes a batch of input sequences.
carry, outputs = nn.RNN(cell, cell_size)(inputs, segmentation_mask)
```
**Bidirectional**
```python
forward_rnn = nn.RNN(nn.LSTMCell(), cell_size=32)
backward_rnn = nn.RNN(nn.LSTMCell(), cell_size=32)
# Bidirectional combinator.
bi_rnn = nn.Bidirectional(forward_rnn, backward_rnn)
# Encodes a batch of input sequences in both directions.
carry, outputs = bi_rnn(inputs, segmentation_mask) 
```
For the complete proposal see #2585.

**Comment by marcvanzee** [comment]
Thank you for this very comprehensive FLIP @cgarciae and @bastings, really impressive how much work you put into this and the proposal is extremely detailed.

### Thoughts / Questions

* **explain current API:** Overall, I think it could be useful to briefly explain the current state of our API. We currently have a single `RNNCellBase` and 4 subclasses `LSTMCell`, `OptimizedLSTMCell`, `GRUCell`, and `ConvLSTM`. Each of these subclasses implements the abstract method `initialize_carry`.... (then explain what this does...)
* **RNN abstraction**: I agree with the proposal that having an RNN abstraction is useful, since it will simplify the masking and scanning logic. In your proposal you suggest adding `RNNBase` and subclass `RNN` (and optionally `GRU`, `LSTM` as well). I have some questions about this:
  * What functionality would be in `LSTM` that is different from an `RNN` with cell `LSTMCell`?
  * If there is no added functionality, and we decide to only add `RNN`: why do we need `RNNBase` in the first place? Can't we just add a single class `RNN`, which is just a recurrent abstraction for one of the cells that can be passed in?
* **make initialize_carry a class method**: I think this is a great idea! Indeed it is a breaking change, but as a first step we could see how many internal users we break and whether it is feasible to update them. Then we could implement a deprecation plan for the old API (similar to what we did for `flax.nn` and `flax.optim`). I really like how it simplifies user code.
* **Bidirectional abstraction**: You propose to add a single class `nn.Bidirectional`. What if we just add an argument `bidirectional` to the `RNN ` abstraction? In that case we might have the limitation that we can only choose the same cell for both the forward and reverse run, but the upshot is that we have one abstraction less to maintain.
* **Masking**: It is not really clear from this proposal how exactly you are going to implement sequence packing. It mentioned some difficulties, but not how you will address them. It was clear to me how you intended to implement sequence length masking (the user provides the sequence lengths), but how does this look for packing? Which inputs should the user provide? How complex will this be?
* **Recurrent dropout:** I am missing a motivation for this. Forgive me my ignorance, but how important is this feature? Is this something that is absolute fundamental to any RNN, or rather a nice-to-have? If it isn't fundamental we could also think of implementing this later if it turns out there is a need for it. The change also seems to involve modifying our `Dropout` layer, which may be fine but it does introduce more complexity.

### Suggestions for implementation plan

One thing that is missing from this (quite big) FLIP is an implementation plan. Which parts are P0s, which are nice-to-haves, how long will each part take?

Thinking about this proposal and my comments above, what do you think of the following minimal (and perhaps sufficient) implementation plan? (Just a suggestion of course 😄 ):
1. Add a single `RNN` abstraction that encapsulates scanning over an `RNNCellBase` instance and add tests.
2. Add masking logic to the `RNN` abstraction. In the proposal you mention sequence packing but I think there are still things unclear here.
3. Add a flag `bidirectional` to the `RNN` abstraction, which runs over the inputs forward and reverse with the same `RNNCellBase` (e.g., `LSTMCell`).
4. Improve the API for `initialize_carry` and deprecate old API.
5. Add recurrent dropout.

Let me know what you think of all this!

**Comment by 8bitmp3** [comment]
Thank you @bastings and @cgarciae 

![image](https://user-images.githubusercontent.com/19637339/187549191-a39db35b-0134-499c-9958-8d450a9ef3b9.png)

**Comment by jheek** [comment]
Thanks for writing this up!

Some thoughts:
- Could you add the benchmark so we can decide whether the pre-computed mask is worth the added complexity (compared to using split_rngs={"dropout": False})
- Please also consider making the carry a proper linnen variable. This would bring cells more in line with the rest of Flax because now carry acts as a manually tracked variable
- I wonder if we can make cells and/or nn.scan flexible enough such that you can just use nn.scan over them directly. Right now RNNBase feels like a special scan primitive for cell.
- I don't see the added benefit of separate RNNBase from RNN so we can have specialized classes for LSTM and GRU. I think other frameworks have had such abstractions in the past mainly because they had specialized kernels which execute specific RNNs faster but we are not constrained by custom kernels.

**Comment by cgarciae** [comment]
FLIP has been moved to a  #2585, feedback was integrated there but will be responding to some of the questions here for completeness:

> **explain current API:** Overall, I think it could be useful to briefly explain the current state of our API. We currently have a single `RNNCellBase` and 4 subclasses `LSTMCell`, `OptimizedLSTMCell`, `GRUCell`, and `ConvLSTM`. Each of these subclasses implements the abstract method `initialize_carry`.... (then explain what this does...)

This explanation was added in the "Motivation" section.

> **RNN abstraction**: I agree with the proposal that having an RNN abstraction is useful, since it will simplify the masking and scanning logic. In your proposal you suggest adding `RNNBase` and subclass `RNN` (and optionally `GRU`, `LSTM` as well). I have some questions about this:
>  * What functionality would be in `LSTM` that is different from an `RNN` with cell `LSTMCell`?
>  * If there is no added functionality, and we decide to only add `RNN`: why do we need `RNNBase` in the first place? 
>
> Can't we just add a single class `RNN`, which is just a recurrent abstraction for one of the cells that can be passed in?
  
We agreed that it was indeed have a single `RNN` abstraction without a Base class. However, it is possible that specialized Modules like e.g. `LSTM` are added if an optimized kernel is provided (see [jax.experimental.rnn](https://github.com/google/jax/blob/main/jax/experimental/rnn.py)).
  
> **make initialize_carry a method**: I think this is a great idea! Indeed it is a breaking change, but as a first step we could see how many internal users we break and whether it is feasible to update them. Then we could implement a deprecation plan for the old API (similar to what we did for `flax.nn` and `flax.optim`). I really like how it simplifies user code.

We chose not to implement any breaking changes, however this improvement could be done in the future along with other breaking changes. @jheek has some ideas how to improve `nn.scan`, some of which could eliminate the need for `initialize_carry` entirely.

> **Bidirectional abstraction**: You propose to add a single class `nn.Bidirectional`. What if we just add an argument `bidirectional` to the `RNN ` abstraction? In that case we might have the limitation that we can only choose the same cell for both the forward and reverse run, but the upshot is that we have one abstraction less to maintain.

We've thought about this, but the implementation would hamper readability which goes against the Flax philosophy. We prefer to keep this in a `Bidirectional` class. 

> **Masking**: It is not really clear from this proposal how exactly you are going to implement sequence packing. It mentioned some difficulties, but not how you will address them. It was clear to me how you intended to implement sequence length masking (the user provides the sequence lengths), but how does this look for packing? Which inputs should the user provide? How complex will this be?

We decided not to implement sequence packing for now, if required this could be implemented in a separate e.g. `PackedRNN` which would have a slightly different carry shape. However, we are keeping the `segmentation_mask` format as its compatible with both packed and 'single sequence per row` implementations.

> **Recurrent dropout:** I am missing a motivation for this. Forgive me my ignorance, but how important is this feature? Is this something that is absolute fundamental to any RNN, or rather a nice-to-have? If it isn't fundamental we could also think of implementing this later if it turns out there is a need for it. The change also seems to involve modifying our `Dropout` layer, which may be fine but it does introduce more complexity.

Ultimately we decided to not go for the Mask APIs (removed from the FLIP) in favor for a simpler strategy which is to leverage Flax's PRNG Streams to differentiate between recurrent dropout and regular dropout. `Dropout` was recently updated to let you specify the RNG stream name. If performance is needed we would recommend users to levarage so of the options provided by the [jax_default_prng_impl](https://jax.readthedocs.io/en/latest/jax.random.html#advanced-rng-configuration) flag.

**Comment by marcvanzee** [comment]
Thanks for the replies @cgarciae! I guess the last point that is unanswered until now are benchmarks. @jheek asked for them above, and @bastings told me that you indeed did some, but I can't find any evidence of this. Could you please add the results in the FLIP or reply here?
